File: backend/app/services/analytics.py
"""Operational analytics.

Every figure here answers a question a public-works supervisor actually asks:
what is coming in, what is stuck, how long do we take, where do we keep going
back, and how are the crews doing. Nothing is computed just to fill a chart.

Aggregation runs in SQL wherever it can. Resolution time is the one exception:
it needs a per-row interval, and computing date differences in portable SQL
across PostgreSQL and SQLite is worse than pulling the small set of resolved
rows and averaging in Python.
"""

from __future__ import annotations

import uuid
from collections import Counter, defaultdict
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.core.enums import (
    CLOSED_STATUSES,
    AssignmentStatus,
    ComplaintStatus,
    PriorityLevel,
)
from app.models.complaint import Complaint
from app.models.repair import RepairAssignment, RepairTeam
from app.repositories.complaint import ComplaintRepository
from app.repositories.system import DuplicateRepository
from app.services.geo import haversine_meters

#: Grid size for "repeat locations" clustering, ~55m.
_CLUSTER_PRECISION = 0.0005


@dataclass(slots=True)
class KpiSummary:
    total_reports: int
    critical: int
    high: int
    in_progress: int
    resolved: int
    pending_review: int
    unresolved: int
    average_resolution_hours: float | None
    resolution_rate: float
    reports_last_7_days: int
    reports_last_30_days: int
    pending_duplicates: int


class AnalyticsService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.complaints = ComplaintRepository(db)
        self.duplicates = DuplicateRepository(db)

    # -- dashboard ------------------------------------------------------

    def kpis(self, now: datetime | None = None) -> KpiSummary:
        now = now or datetime.now(UTC)
        by_status = self.complaints.count_by_status()
        by_level = self.complaints.count_by_priority_level()

        total = self.complaints.total()
        resolved = by_status.get(ComplaintStatus.RESOLVED.value, 0)
        closed = sum(by_status.get(status.value, 0) for status in CLOSED_STATUSES)

        return KpiSummary(
            total_reports=total,
            critical=by_level.get(PriorityLevel.CRITICAL.value, 0),
            high=by_level.get(PriorityLevel.HIGH.value, 0),
            in_progress=by_status.get(ComplaintStatus.IN_PROGRESS.value, 0),
            resolved=resolved,
            pending_review=(
                by_status.get(ComplaintStatus.PENDING.value, 0)
                + by_status.get(ComplaintStatus.AI_ANALYZED.value, 0)
                + by_status.get(ComplaintStatus.PRIORITIZED.value, 0)
            ),
            unresolved=total - closed,
            average_resolution_hours=self.average_resolution_hours(),
            # Rate over *actioned* reports: counting rejected and duplicate rows
            # as failures would understate a team that triaged them correctly.
            resolution_rate=round(resolved / total * 100, 1) if total else 0.0,
            reports_last_7_days=self.complaints.count_created_since(now - timedelta(days=7)),
            reports_last_30_days=self.complaints.count_created_since(now - timedelta(days=30)),
            pending_duplicates=self.duplicates.pending_count(),
        )

    def average_resolution_hours(self) -> float | None:
        rows = self.db.execute(
            select(Complaint.created_at, Complaint.resolved_at).where(
                Complaint.resolved_at.is_not(None)
            )
        ).all()
        if not rows:
            return None
        durations = [
            (_utc(resolved) - _utc(created)).total_seconds() / 3600
            for created, resolved in rows
            if resolved and created
        ]
        durations = [value for value in durations if value >= 0]
        if not durations:
            return None
        return round(sum(durations) / len(durations), 1)

    # -- time series ----------------------------------------------------

    def reports_over_time(self, days: int = 30, now: datetime | None = None) -> list[dict]:
        """Daily counts, gap-filled so the chart has no missing days."""
        now = now or datetime.now(UTC)
        start = (now - timedelta(days=days - 1)).replace(hour=0, minute=0, second=0, microsecond=0)

        rows = self.db.execute(
            select(Complaint.created_at, Complaint.status, Complaint.priority_level).where(
                Complaint.created_at >= start
            )
        ).all()

        created_by_day: Counter[str] = Counter()
        critical_by_day: Counter[str] = Counter()
        for created_at, _status, level in rows:
            key = _utc(created_at).date().isoformat()
            created_by_day[key] += 1
            if level == PriorityLevel.CRITICAL:
                critical_by_day[key] += 1

        resolved_rows = self.db.execute(
            select(Complaint.resolved_at).where(Complaint.resolved_at >= start)
        ).all()
        resolved_by_day: Counter[str] = Counter()
        for (resolved_at,) in resolved_rows:
            if resolved_at:
                resolved_by_day[_utc(resolved_at).date().isoformat()] += 1

        series = []
        for offset in range(days):
            day = (start + timedelta(days=offset)).date().isoformat()
            series.append(
                {
                    "date": day,
                    "reported": created_by_day.get(day, 0),
                    "resolved": resolved_by_day.get(day, 0),
                    "critical": critical_by_day.get(day, 0),
                }
            )
        return series

    def by_damage_type(self) -> list[dict]:
        counts = self.complaints.count_by_damage_type()
        return [
            {"damage_type": key, "count": value}
            for key, value in sorted(counts.items(), key=lambda item: item[1], reverse=True)
        ]

    def priority_distribution(self) -> list[dict]:
        counts = self.complaints.count_by_priority_level()
        return [
            {"level": level.value, "count": counts.get(level.value, 0)}
            for level in (
                PriorityLevel.CRITICAL,
                PriorityLevel.HIGH,
                PriorityLevel.MEDIUM,
                PriorityLevel.LOW,
            )
        ]

    def status_distribution(self) -> list[dict]:
        counts = self.complaints.count_by_status()
        return [{"status": status.value, "count": counts.get(status.value, 0)} for status in ComplaintStatus]

    def by_area(self, limit: int = 10) -> list[dict]:
        """Volume by administrative area, from the resolved location record."""
        from app.models.geo import Location

        rows = self.db.execute(
            select(
                func.coalesce(Location.city, "Unknown"),
                func.count(Complaint.id),
                func.avg(Complaint.priority_score),
            )
            .join(Complaint, Complaint.id == Location.complaint_id)
            .group_by(func.coalesce(Location.city, "Unknown"))
            .order_by(func.count(Complaint.id).desc())
            .limit(limit)
        ).all()
        return [
            {"area": area, "count": int(count), "average_priority": round(float(avg or 0), 1)}
            for area, count, avg in rows
        ]

    def top_roads(self, limit: int = 10) -> list[dict]:
        """Roads generating the most reports - the recurring-failure list."""
        rows = self.db.execute(
            select(
                Complaint.road_name,
                func.count(Complaint.id),
                func.avg(Complaint.priority_score),
                func.max(Complaint.priority_score),
            )
            .where(Complaint.road_name.is_not(None), Complaint.road_name != "")
            .group_by(Complaint.road_name)
            .order_by(func.count(Complaint.id).desc(), func.avg(Complaint.priority_score).desc())
            .limit(limit)
        ).all()
        return [
            {
                "road_name": road,
                "count": int(count),
                "average_priority": round(float(avg or 0), 1),
                "max_priority": round(float(peak or 0), 1),
            }
            for road, count, avg, peak in rows
        ]

    def repeat_locations(self, limit: int = 10, min_reports: int = 2) -> list[dict]:
        """Spots reported more than once - where repairs are not holding.

        Coordinates are snapped to a ~55m grid, then neighbouring cells are
        merged, so two reports either side of a grid boundary still cluster.
        """
        rows = self.db.execute(
            select(
                Complaint.latitude,
                Complaint.longitude,
                Complaint.road_name,
                Complaint.priority_score,
                Complaint.status,
            )
        ).all()

        buckets: dict[tuple[int, int], list] = defaultdict(list)
        for latitude, longitude, road, score, status in rows:
            key = (round(latitude / _CLUSTER_PRECISION), round(longitude / _CLUSTER_PRECISION))
            buckets[key].append((latitude, longitude, road, score, status))

        clusters = []
        for entries in buckets.values():
            if len(entries) < min_reports:
                continue
            latitudes = [item[0] for item in entries]
            longitudes = [item[1] for item in entries]
            centre_lat = sum(latitudes) / len(latitudes)
            centre_lon = sum(longitudes) / len(longitudes)
            spread = max(
                haversine_meters(centre_lat, centre_lon, lat, lon)
                for lat, lon in zip(latitudes, longitudes, strict=True)
            )
            roads = [item[2] for item in entries if item[2]]
            unresolved = sum(1 for item in entries if item[4] not in CLOSED_STATUSES)
            clusters.append(
                {
                    "latitude": round(centre_lat, 6),
                    "longitude": round(centre_lon, 6),
                    "road_name": Counter(roads).most_common(1)[0][0] if roads else None,
                    "report_count": len(entries),
                    "unresolved": unresolved,
                    "max_priority": round(max(item[3] for item in entries), 1),
                    "spread_meters": round(spread, 1),
                }
            )

        clusters.sort(key=lambda item: (item["report_count"], item["max_priority"]), reverse=True)
        return clusters[:limit]

    # -- teams ----------------------------------------------------------

    def team_performance(self) -> list[dict]:
        teams = self.db.execute(select(RepairTeam).order_by(RepairTeam.name)).scalars().all()
        assignments = self.db.execute(
            select(RepairAssignment).options()
        ).scalars().all()

        by_team: dict[uuid.UUID, list[RepairAssignment]] = defaultdict(list)
        for assignment in assignments:
            by_team[assignment.team_id].append(assignment)

        results = []
        for team in teams:
            items = by_team.get(team.id, [])
            completed = [
                item
                for item in items
                if item.status in (AssignmentStatus.COMPLETED, AssignmentStatus.VERIFIED)
            ]
            open_jobs = [
                item
                for item in items
                if item.status in (AssignmentStatus.ASSIGNED, AssignmentStatus.IN_PROGRESS)
            ]

            durations = [
                (_utc(item.completed_at) - _utc(item.created_at)).total_seconds() / 3600
                for item in completed
                if item.completed_at and item.created_at
            ]
            on_time = [
                item
                for item in completed
                if item.due_at and item.completed_at and _utc(item.completed_at) <= _utc(item.due_at)
            ]

            results.append(
                {
                    "team_id": str(team.id),
                    "team_name": team.name,
                    "zone": team.zone,
                    "total_assigned": len(items),
                    "completed": len(completed),
                    "open_jobs": len(open_jobs),
                    "capacity": team.max_concurrent_jobs,
                    "average_completion_hours": (
                        round(sum(durations) / len(durations), 1) if durations else None
                    ),
                    "on_time_rate": (
                        round(len(on_time) / len(completed) * 100, 1) if completed else None
                    ),
                }
            )

        results.sort(key=lambda item: item["completed"], reverse=True)
        return results

    def resolution_by_priority(self) -> list[dict]:
        """Are the urgent ones actually being done first?"""
        rows = self.db.execute(
            select(Complaint.priority_level, Complaint.created_at, Complaint.resolved_at).where(
                Complaint.resolved_at.is_not(None)
            )
        ).all()

        grouped: dict[str, list[float]] = defaultdict(list)
        for level, created, resolved in rows:
            if created and resolved:
                grouped[str(level)].append((_utc(resolved) - _utc(created)).total_seconds() / 3600)

        return [
            {
                "level": level.value,
                "resolved_count": len(grouped.get(level.value, [])),
                "average_hours": (
                    round(sum(grouped[level.value]) / len(grouped[level.value]), 1)
                    if grouped.get(level.value)
                    else None
                ),
            }
            for level in (
                PriorityLevel.CRITICAL,
                PriorityLevel.HIGH,
                PriorityLevel.MEDIUM,
                PriorityLevel.LOW,
            )
        ]


def _utc(value: datetime) -> datetime:
    return value if value.tzinfo else value.replace(tzinfo=UTC)

File: supabase/functions/api/services/analytics.ts
// Ported from backend/app/services/analytics.py. Only kpis() so far (needed
// by the public /stats/public endpoint); the admin-only breakdowns
// (reports_over_time, by_damage_type, repeat_locations, team_performance,
// ...) are ported alongside the admin analytics routes.
//
// Always called with the service client: these are aggregate counts over
// every complaint regardless of status, the same unrestricted view Python's
// single DB connection had - RLS would otherwise silently undercount by
// excluding REJECTED/DUPLICATE rows from a citizen or anon caller's view.

import type { SupabaseClient } from "@supabase/supabase-js";
import type { ComplaintStatus, PriorityLevel } from "../_shared/enums.ts";
import { CLOSED_STATUSES } from "../_shared/enums.ts";
import { haversineMeters } from "./geo.ts";

export interface KpiSummary {
  totalReports: number;
  critical: number;
  high: number;
  inProgress: number;
  resolved: number;
  pendingReview: number;
  unresolved: number;
  averageResolutionHours: number | null;
  resolutionRate: number;
  reportsLast7Days: number;
  reportsLast30Days: number;
  pendingDuplicates: number;
  /** Repairs a crew has marked done that no admin has approved yet. */
  awaitingVerification: number;
  /** Reports whose active repair has passed its due date. */
  overdue: number;
}

function round(value: number, decimals: number): number {
  const f = 10 ** decimals;
  return Math.round(value * f) / f;
}

/** What the kpi_summary() database function returns; see its migration. */
interface KpiRow {
  total: number;
  by_status: Record<string, number>;
  by_priority_level: Record<string, number>;
  average_resolution_hours: number | null;
  last_7_days: number;
  last_30_days: number;
  pending_duplicates: number;
  awaiting_verification: number;
  overdue: number;
}

/**
 * Every headline figure and both distributions come from one database call.
 * The counting used to happen here, after downloading every complaint row -
 * slow, growing with the table, and silently wrong past PostgREST's 1000-row
 * cap. See supabase/migrations/20260922020000_kpi_summary_rpc.sql.
 */
async function kpiRow(client: SupabaseClient): Promise<KpiRow> {
  const { data, error } = await client.rpc("kpi_summary");
  if (error) throw error;
  return data as KpiRow;
}

export async function kpis(client: SupabaseClient): Promise<KpiSummary> {
  const row = await kpiRow(client);
  const byStatus = row.by_status;

  const total = row.total;
  const resolved = byStatus["RESOLVED"] ?? 0;
  const closed = (CLOSED_STATUSES as ComplaintStatus[]).reduce((sum, s) => sum + (byStatus[s] ?? 0), 0);

  return {
    totalReports: total,
    critical: row.by_priority_level["CRITICAL" as PriorityLevel] ?? 0,
    high: row.by_priority_level["HIGH" as PriorityLevel] ?? 0,
    inProgress: byStatus["IN_PROGRESS"] ?? 0,
    resolved,
    pendingReview: (byStatus["PENDING"] ?? 0) + (byStatus["AI_ANALYZED"] ?? 0) + (byStatus["PRIORITIZED"] ?? 0),
    unresolved: total - closed,
    averageResolutionHours: row.average_resolution_hours,
    resolutionRate: total ? round((resolved / total) * 100, 1) : 0.0,
    reportsLast7Days: row.last_7_days,
    reportsLast30Days: row.last_30_days,
    pendingDuplicates: row.pending_duplicates,
    awaitingVerification: row.awaiting_verification,
    overdue: row.overdue,
  };
}

// -- the admin-only breakdowns (services/analytics.py's remaining methods) --

const PRIORITY_LEVELS_DESC: PriorityLevel[] = ["CRITICAL", "HIGH", "MEDIUM", "LOW"];
const ALL_STATUSES: ComplaintStatus[] = [
  "PENDING", "AI_ANALYZED", "PRIORITIZED", "ASSIGNED", "IN_PROGRESS", "RESOLVED", "REJECTED", "DUPLICATE",
];

/** Daily counts for the last `days` days, gap-filled so the chart has no missing days. */
export async function reportsOverTime(client: SupabaseClient, days = 30, now: Date = new Date()): Promise<Record<string, unknown>[]> {
  const start = new Date(now.getTime() - (days - 1) * 86_400_000);
  start.setUTCHours(0, 0, 0, 0);

  const { data: created, error: createdErr } = await client
    .from("complaints")
    .select("created_at, priority_level")
    .gte("created_at", start.toISOString());
  if (createdErr) throw createdErr;

  const { data: resolved, error: resolvedErr } = await client
    .from("complaints")
    .select("resolved_at")
    .not("resolved_at", "is", null)
    .gte("resolved_at", start.toISOString());
  if (resolvedErr) throw resolvedErr;

  const dayKey = (iso: string) => iso.slice(0, 10);
  const createdByDay: Record<string, number> = {};
  const criticalByDay: Record<string, number> = {};
  for (const row of created ?? []) {
    const key = dayKey(row.created_at);
    createdByDay[key] = (createdByDay[key] ?? 0) + 1;
    if (row.priority_level === "CRITICAL") criticalByDay[key] = (criticalByDay[key] ?? 0) + 1;
  }
  const resolvedByDay: Record<string, number> = {};
  for (const row of resolved ?? []) {
    const key = dayKey(row.resolved_at);
    resolvedByDay[key] = (resolvedByDay[key] ?? 0) + 1;
  }

  const series: Record<string, unknown>[] = [];
  for (let offset = 0; offset < days; offset++) {
    const day = new Date(start.getTime() + offset * 86_400_000).toISOString().slice(0, 10);
    series.push({
      date: day,
      reported: createdByDay[day] ?? 0,
      resolved: resolvedByDay[day] ?? 0,
      critical: criticalByDay[day] ?? 0,
    });
  }
  return series;
}

export async function byDamageType(client: SupabaseClient): Promise<Record<string, unknown>[]> {
  const { data, error } = await client.from("complaints").select("damage_type");
  if (error) throw error;
  const counts: Record<string, number> = {};
  for (const row of data ?? []) counts[row.damage_type] = (counts[row.damage_type] ?? 0) + 1;
  return Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .map(([damage_type, count]) => ({ damage_type, count }));
}

export async function priorityDistribution(client: SupabaseClient): Promise<Record<string, unknown>[]> {
  const counts = (await kpiRow(client)).by_priority_level;
  return PRIORITY_LEVELS_DESC.map((level) => ({ level, count: counts[level] ?? 0 }));
}

export async function statusDistribution(client: SupabaseClient): Promise<Record<string, unknown>[]> {
  const counts = (await kpiRow(client)).by_status;
  return ALL_STATUSES.map((status) => ({ status, count: counts[status] ?? 0 }));
}

/** Volume by administrative area, from the resolved location record. */
export async function byArea(client: SupabaseClient, limit = 10): Promise<Record<string, unknown>[]> {
  const { data, error } = await client.from("locations").select("city, complaint_id, complaints(priority_score)");
  if (error) throw error;

  const grouped: Record<string, { count: number; totalPriority: number }> = {};
  for (const row of (data ?? []) as any[]) {
    const area = row.city ?? "Unknown";
    const bucket = grouped[area] ?? { count: 0, totalPriority: 0 };
    bucket.count += 1;
    bucket.totalPriority += Number(row.complaints?.priority_score ?? 0);
    grouped[area] = bucket;
  }

  return Object.entries(grouped)
    .map(([area, { count, totalPriority }]) => ({ area, count, average_priority: round(totalPriority / count, 1) }))
    .sort((a, b) => (b.count as number) - (a.count as number))
    .slice(0, limit);
}

/** Roads generating the most reports - the recurring-failure list. */
export async function topRoads(client: SupabaseClient, limit = 10): Promise<Record<string, unknown>[]> {
  const { data, error } = await client.from("complaints").select("road_name, priority_score").not("road_name", "is", null).neq("road_name", "");
  if (error) throw error;

  const grouped: Record<string, number[]> = {};
  for (const row of data ?? []) {
    (grouped[row.road_name] ??= []).push(row.priority_score);
  }

  return Object.entries(grouped)
    .map(([road_name, scores]) => ({
      road_name,
      count: scores.length,
      average_priority: round(scores.reduce((a, b) => a + b, 0) / scores.length, 1),
      max_priority: round(Math.max(...scores), 1),
    }))
    .sort((a, b) => (b.count as number) - (a.count as number) || (b.average_priority as number) - (a.average_priority as number))
    .slice(0, limit);
}

const CLUSTER_PRECISION = 0.0005;

/**
 * Spots reported more than once - where repairs are not holding. Coordinates
 * are snapped to a ~55m grid, then neighbouring cells are merged, so two
 * reports either side of a grid boundary still cluster.
 */
export async function repeatLocations(client: SupabaseClient, limit = 10, minReports = 2): Promise<Record<string, unknown>[]> {
  const { data, error } = await client.from("complaints").select("latitude, longitude, road_name, priority_score, status");
  if (error) throw error;

  const buckets = new Map<string, { latitude: number; longitude: number; road_name: string | null; priority_score: number; status: string }[]>();
  for (const row of data ?? []) {
    const key = `${Math.round(row.latitude / CLUSTER_PRECISION)}:${Math.round(row.longitude / CLUSTER_PRECISION)}`;
    (buckets.get(key) ?? buckets.set(key, []).get(key)!).push(row);
  }

  const clusters: Record<string, unknown>[] = [];
  for (const entries of buckets.values()) {
    if (entries.length < minReports) continue;
    const centreLat = entries.reduce((s, e) => s + e.latitude, 0) / entries.length;
    const centreLon = entries.reduce((s, e) => s + e.longitude, 0) / entries.length;
    const spread = Math.max(...entries.map((e) => haversineMeters(centreLat, centreLon, e.latitude, e.longitude)));
    const roadCounts: Record<string, number> = {};
    for (const e of entries) if (e.road_name) roadCounts[e.road_name] = (roadCounts[e.road_name] ?? 0) + 1;
    const topRoad = Object.entries(roadCounts).sort((a, b) => b[1] - a[1])[0]?.[0] ?? null;
    const unresolved = entries.filter((e) => !(CLOSED_STATUSES as string[]).includes(e.status)).length;

    clusters.push({
      latitude: round(centreLat, 6),
      longitude: round(centreLon, 6),
      road_name: topRoad,
      report_count: entries.length,
      unresolved,
      max_priority: round(Math.max(...entries.map((e) => e.priority_score)), 1),
      spread_meters: round(spread, 1),
    });
  }

  clusters.sort((a, b) => (b.report_count as number) - (a.report_count as number) || (b.max_priority as number) - (a.max_priority as number));
  return clusters.slice(0, limit);
}

export async function teamPerformance(client: SupabaseClient): Promise<Record<string, unknown>[]> {
  const { data: teams, error: teamsErr } = await client.from("repair_teams").select("*").order("name", { ascending: true });
  if (teamsErr) throw teamsErr;
  const { data: assignments, error: assignmentsErr } = await client.from("repair_assignments").select("*");
  if (assignmentsErr) throw assignmentsErr;

  const byTeam = new Map<string, any[]>();
  for (const a of assignments ?? []) (byTeam.get(a.team_id) ?? byTeam.set(a.team_id, []).get(a.team_id)!).push(a);

  const results = (teams ?? []).map((team) => {
    const items = byTeam.get(team.id) ?? [];
    const completed = items.filter((i) => i.status === "COMPLETED" || i.status === "VERIFIED");
    const openJobs = items.filter((i) => i.status === "ASSIGNED" || i.status === "IN_PROGRESS");

    const durations = completed
      .filter((i) => i.completed_at && i.created_at)
      .map((i) => (new Date(i.completed_at).getTime() - new Date(i.created_at).getTime()) / 3_600_000);
    const onTime = completed.filter((i) => i.due_at && i.completed_at && new Date(i.completed_at) <= new Date(i.due_at));
    // Every assignment that was ever sent back at least once, whether or not
    // it has since been verified - a job resubmitted after rework still
    // counts, same as `completed` above counts it once done regardless of
    // how it got there.
    const reworked = items.filter((i) => (i.rework_count ?? 0) > 0);

    return {
      team_id: team.id,
      team_name: team.name,
      zone: team.zone,
      total_assigned: items.length,
      completed: completed.length,
      open_jobs: openJobs.length,
      capacity: team.max_concurrent_jobs,
      average_completion_hours: durations.length ? round(durations.reduce((a, b) => a + b, 0) / durations.length, 1) : null,
      on_time_rate: completed.length ? round((onTime.length / completed.length) * 100, 1) : null,
      rework_rate: completed.length ? round((reworked.length / completed.length) * 100, 1) : null,
    };
  });

  results.sort((a, b) => (b.completed as number) - (a.completed as number));
  return results;
}

/** Are the urgent ones actually being done first? */
export async function resolutionByPriority(client: SupabaseClient): Promise<Record<string, unknown>[]> {
  const { data, error } = await client
    .from("complaints")
    .select("priority_level, created_at, resolved_at")
    .not("resolved_at", "is", null);
  if (error) throw error;

  const grouped: Record<string, number[]> = {};
  for (const row of data ?? []) {
    if (row.created_at && row.resolved_at) {
      (grouped[row.priority_level] ??= []).push((new Date(row.resolved_at).getTime() - new Date(row.created_at).getTime()) / 3_600_000);
    }
  }

  return PRIORITY_LEVELS_DESC.map((level) => {
    const hours = grouped[level] ?? [];
    return {
      level,
      resolved_count: hours.length,
      average_hours: hours.length ? round(hours.reduce((a, b) => a + b, 0) / hours.length, 1) : null,
    };
  });
}
